File: src/dep_health_scanner/reporter.py
from __future__ import annotations

from collections import defaultdict
from typing import List

from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from rich import box

from .models import ScanResult


class Reporter:
    def __init__(self, fail_on_critical: bool = False):
        self.console = Console()
        self.fail_on_critical = fail_on_critical

    def print_report(self, results: List[ScanResult]) -> int:
        if not results:
            self.console.print("[yellow]No dependencies scanned.[/yellow]")
            return 0

        # Classify
        buckets = defaultdict(list)
        for r in results:
            level = self._classify(r)
            buckets[level].append(r)

        self.console.print()
        self.console.print(
            Panel.fit(
                "[bold cyan]DEPENDENCY HEALTH REPORT[/bold cyan]",
                border_style="cyan",
            )
        )
        self.console.print()

        if "critical" in buckets:
            self._print_section("CRITICAL ISSUES", "bold red", buckets["critical"])

        if "high" in buckets:
            self._print_section("HIGH SEVERITY", "bold red", buckets["high"])

        if "medium" in buckets:
            self._print_section("MEDIUM SEVERITY", "bold yellow", buckets["medium"])

        if "low" in buckets:
            self._print_section("LOW SEVERITY", "bold yellow", buckets["low"])

        # Summary
        table = Table(
            title="SUMMARY",
            box=box.SIMPLE,
            show_header=True,
            header_style="bold cyan",
        )
        table.add_column("Metric", style="bold")
        table.add_column("Count", justify="right")
        table.add_row("Total dependencies", str(len(results)))
        table.add_row(
            "Vulnerable",
            str(sum(1 for r in results if r.vulnerabilities)),
            style="red",
        )
        table.add_row(
            "Outdated",
            str(sum(1 for r in results if r.outdated)),
            style="yellow",
        )
        table.add_row(
            "Critical",
            str(len(buckets.get("critical", []))),
            style="red bold",
        )
        table.add_row(
            "High",
            str(len(buckets.get("high", []))),
            style="red",
        )
        table.add_row(
            "Medium",
            str(len(buckets.get("medium", []))),
            style="yellow",
        )
        table.add_row(
            "Low",
            str(len(buckets.get("low", []))),
            style="green",
        )
        self.console.print(table)
        self.console.print()

        critical_count = len(buckets.get("critical", []))
        if self.fail_on_critical and critical_count > 0:
            return 1
        return 0

    def _print_section(self, title: str, style: str, results: List[ScanResult]):
        self.console.print(f"[{style}]{title}[/{style}]")
        self.console.print()
        for r in results:
            icon = "✖" if "CRITICAL" in style or "HIGH" in style else "⚠"
            self.console.print(
                f"  {icon} [cyan]{r.dependency.name}[/cyan]@[dim]{r.dependency.version}[/dim] [{r.dependency.ecosystem.value}]"
            )
            if r.latest_version:
                self.console.print(
                    f"       [dim]latest:[/dim] [green]{r.latest_version}[/green]"
                )
            for v in r.vulnerabilities:
                sev = (
                    "CRITICAL"
                    if v.severity >= 9.0
                    else "HIGH"
                    if v.severity >= 7.0
                    else "MEDIUM"
                    if v.severity >= 4.0
                    else "LOW"
                )
                color = "red" if sev in ("CRITICAL", "HIGH") else "yellow"
                self.console.print(
                    f"       [red bold]vulnerability:[/red bold] [cyan]{v.id}[/cyan] [{color}]{sev}[/color]: {v.summary}"
                )
                if v.fixed_versions:
                    self.console.print(
                        f"              [green]fixed in:[/green] {', '.join(v.fixed_versions)}"
                    )
        self.console.print()

    def _classify(self, r: ScanResult) -> str:
        if r.vulnerabilities:
            for v in r.vulnerabilities:
                if v.severity >= 9.0:
                    return "critical"
                if v.severity >= 7.0:
                    return "high"
            return "medium"
        if r.outdated:
            return "low"
        return "ok"

File: src/reporter.rs
use super::dependency::ScanResult;
use colored::Colorize;
use std::collections::HashMap;

pub struct Reporter {
    fail_on_critical: bool,
    #[allow(dead_code)]
    min_severity: String,
}

impl Reporter {
    pub fn new(fail_on_critical: bool, min_severity: String) -> Self {
        Self {
            fail_on_critical,
            min_severity,
        }
    }

    pub fn print_report(&self, results: &[ScanResult]) -> crate::Result<()> {
        if results.is_empty() {
            println!("{}", "No dependencies scanned.".yellow());
            return Ok(());
        }

        let total = results.len();
        let mut count_by_severity = HashMap::new();
        count_by_severity.insert("critical", 0);
        count_by_severity.insert("high", 0);
        count_by_severity.insert("medium", 0);
        count_by_severity.insert("low", 0);
        count_by_severity.insert("vulnerable", 0);
        count_by_severity.insert("outdated", 0);

        let mut critical_found = false;

        // Summary header
        println!(
            "\n{}",
            "═══════════════════════════════════════".cyan().bold()
        );
        println!(
            "{}",
            "       DEPENDENCY HEALTH REPORT        ".cyan().bold()
        );
        println!(
            "{}",
            "═══════════════════════════════════════".cyan().bold()
        );
        println!();

        // Group results by severity
        let mut critical = Vec::new();
        let mut high = Vec::new();
        let mut medium = Vec::new();
        let mut low = Vec::new();
        let mut ok = Vec::new();

        for result in results {
            let severity = self.classify_result(result);
            match severity {
                "critical" => {
                    critical_found = true;
                    critical.push(result);
                    *count_by_severity.get_mut("critical").unwrap() += 1;
                    *count_by_severity.get_mut("vulnerable").unwrap() += 1;
                }
                "high" => {
                    high.push(result);
                    *count_by_severity.get_mut("high").unwrap() += 1;
                    if !result.vulnerabilities.is_empty() {
                        *count_by_severity.get_mut("vulnerable").unwrap() += 1;
                    }
                    if result.outdated {
                        *count_by_severity.get_mut("outdated").unwrap() += 1;
                    }
                }
                "medium" => {
                    medium.push(result);
                    *count_by_severity.get_mut("medium").unwrap() += 1;
                    if result.outdated {
                        *count_by_severity.get_mut("outdated").unwrap() += 1;
                    }
                }
                "low" => {
                    low.push(result);
                    *count_by_severity.get_mut("low").unwrap() += 1;
                    if result.outdated {
                        *count_by_severity.get_mut("outdated").unwrap() += 1;
                    }
                }
                _ => {
                    ok.push(result);
                }
            }
        }

        // Print critical issues first
        if !critical.is_empty() {
            println!("{}", "CRITICAL ISSUES:".red().bold());
            println!();
            for result in &critical {
                self.print_dependency(result, "critical");
            }
            println!();
        }

        // Print high severity
        if !high.is_empty() {
            println!("{}", "HIGH SEVERITY:".red());
            println!();
            for result in &high {
                self.print_dependency(result, "high");
            }
            println!();
        }

        // Print medium severity
        if !medium.is_empty() {
            println!("{}", "MEDIUM SEVERITY:".yellow());
            println!();
            for result in &medium {
                self.print_dependency(result, "medium");
            }
            println!();
        }

        // Print low severity
        if !low.is_empty() {
            println!("{}", "LOW SEVERITY:".yellow());
            println!();
            for result in &low {
                self.print_dependency(result, "low");
            }
            println!();
        }

        // Print summary table
        println!("{}", "───────────────────────────────────────".cyan());
        println!("{}", "SUMMARY".cyan().bold());
        println!("{}", "───────────────────────────────────────".cyan());
        println!(
            "{:<20} {:>6}",
            "Total dependencies:".bold(),
            total.to_string().cyan()
        );
        println!(
            "{:<20} {:>6}",
            "Vulnerable:".bold(),
            count_by_severity["vulnerable"].to_string().red()
        );
        println!(
            "{:<20} {:>6}",
            "Outdated:".bold(),
            count_by_severity["outdated"].to_string().yellow()
        );
        println!(
            "{:<20} {:>6} / {:>6} / {:>6} / {:>6}",
            "By severity:".bold(),
            count_by_severity["critical"].to_string().red(),
            count_by_severity["high"].to_string().red(),
            count_by_severity["medium"].to_string().yellow(),
            count_by_severity["low"].to_string().green()
        );
        println!();

        if self.fail_on_critical && critical_found {
            std::process::exit(1);
        }

        Ok(())
    }

    fn print_dependency(&self, result: &ScanResult, severity: &str) {
        let icon = match severity {
            "critical" => "✖".red().bold(),
            "high" => "✖".red(),
            "medium" => "⚠".yellow(),
            "low" => "ℹ".blue(),
            _ => "✓".green(),
        };

        println!(
            "  {} {}@{} [{}]",
            icon,
            result.dependency.name.cyan(),
            result.dependency.version.dimmed(),
            result.dependency.ecosystem
        );

        if let Some(ref latest) = result.latest_version {
            println!(
                "       {} {} → {}",
                "latest:".dimmed(),
                result.dependency.version.dimmed(),
                latest.green()
            );
        }

        for vuln in &result.vulnerabilities {
            let sev = if vuln.severity >= 9.0 {
                "CRITICAL".red().bold()
            } else if vuln.severity >= 7.0 {
                "HIGH".red()
            } else if vuln.severity >= 4.0 {
                "MEDIUM".yellow()
            } else {
                "LOW".blue()
            };

            println!(
                "       {} {} ({}): {}",
                "vulnerability:".red().bold(),
                vuln.id.cyan(),
                sev,
                vuln.summary
            );

            if !vuln.fixed_versions.is_empty() {
                println!(
                    "              {} in {}",
                    "fixed:".green(),
                    vuln.fixed_versions.join(", ").green()
                );
            }
        }

        if let Some(score) = result.bus_factor_score {
            let score_str = if score < 30 {
                format!("{} (low)", score).red()
            } else if score < 60 {
                format!("{} (medium)", score).yellow()
            } else {
                format!("{} (good)", score).green()
            };
            println!("       {} {}", "bus factor:".dimmed(), score_str);
        }

        if !result.alternatives.is_empty() {
            println!("       {}:", "alternatives".dimmed());
            for alt in &result.alternatives {
                println!(
                    "         - {}@{} (stars: {}, downloads: {})",
                    alt.name.cyan(),
                    alt.version.dimmed(),
                    alt.stars,
                    alt.weekly_downloads
                );
            }
            println!(
                "         {}",
                "run `depscan suggest <package>` for details"
                    .italic()
                    .dimmed()
            );
        }

        println!();
    }

    fn classify_result(&self, result: &ScanResult) -> &'static str {
        if !result.vulnerabilities.is_empty() {
            for vuln in &result.vulnerabilities {
                if vuln.severity >= 9.0 {
                    return "critical";
                }
                if vuln.severity >= 7.0 {
                    return "high";
                }
            }
            return "medium";
        }

        if let Some(score) = result.bus_factor_score {
            if score < 30 {
                return "high";
            }
        }

        if result.outdated {
            return "low";
        }

        "ok"
    }
}
